src/shared/dissect-image.c

```diff
@@ -1409,56 +1414,101 @@ int decrypted_image_relinquish(DecryptedImage *d) {
 
         if (is_device_path(image)) {
                 /* If we are asked to load the root hash for a device node, exit early */
-                *ret = NULL;
-                *ret_size = 0;
+                if (roothash)
+                        *roothash = NULL;
+                if (roothash_size)
+                        *roothash_size = 0;
+                if (hashdevice)
+                        *hashdevice = NULL;
                 return 0;
         }
 
-        r = getxattr_malloc(image, "user.verity.roothash", &text, true);
-        if (r < 0) {
-                char *fn, *e, *n;
+        if (roothash) {
+                assert(roothash_size);
 
-                if (!IN_SET(r, -ENODATA, -EOPNOTSUPP, -ENOENT))
-                        return r;
+                r = getxattr_malloc(image, "user.verity.roothash", &text, true);
+                if (r < 0) {
+                        char *fn, *e, *n;
 
-                fn = newa(char, strlen(image) + STRLEN(".roothash") + 1);
-                n = stpcpy(fn, image);
-                e = endswith(fn, ".raw");
-                if (e)
-                        n = e;
+                        if (!IN_SET(r, -ENODATA, -EOPNOTSUPP, -ENOENT))
+                                return r;
 
-                strcpy(n, ".roothash");
+                        fn = newa(char, strlen(image) + STRLEN(".roothash") + 1);
+                        n = stpcpy(fn, image);
+                        e = endswith(fn, ".raw");
+                        if (e)
+                                n = e;
+                        else {
+                                e = endswith(fn, ".squashfs");
+                                if (e)
+                                        n = e;
+                        }
 
-                r = read_one_line_file(fn, &text);
-                if (r == -ENOENT) {
-                        *ret = NULL;
-                        *ret_size = 0;
-                        return 0;
+                        strcpy(n, ".roothash");
+
+                        r = read_one_line_file(fn, &text);
+                        if (r == -ENOENT) {
+                                *roothash = NULL;
+                                *roothash_size = 0;
+                                return 0;
+                        }
+                        if (r < 0)
+                                return r;
                 }
+
+                r = unhexmem(text, strlen(text), &k, &l);
                 if (r < 0)
                         return r;
+                if (l < sizeof(sd_id128_t))
+                        return -EINVAL;
+
+                *roothash = TAKE_PTR(k);
+                *roothash_size = l;
         }
 
-        r = unhexmem(text, strlen(text), &k, &l);
-        if (r < 0)
-                return r;
-        if (l < sizeof(sd_id128_t))
-                return -EINVAL;
+        if (hashdevice) {
+                r = getxattr_malloc(image, "user.verity.hashdevice", hashdevice, true);
```

Comment:
I picked .hashdevice as the extension to match util-linux's command line option names, to have some cross-tool consistency for handling the same objects. But if there's a preference for a different naming scheme I'm ok with changing to something else.

Reply:
hmm, i wonder if this should be an xattr at all... i mean, you are looking for a file here, and hence might as well look for that right-away. i.e. the indirection via an xattr seems unnecessary. This is different for the roothash which itself isn't a path but metadata directly associated with the image where an xattr makes a lot more sense

Reply:
Sure, was just following the same pattern. Will remove.